Direct Chunk Write

GDYendell opened the pull request:
This is an initial working version of direct chunk write support. I have tested with data direct from ADSimDetector and from NDPlugin Codec. 

The logic is as follows:
- HDF5 plugin settings match the incoming NDArrays (compressed or otherwise) - Use direct chunk write
- HDF5 chunking does not match, or compression configured and array not already compressed - Use standard write
- Data is compressed but does not match the HDF5 plugin (compression settings or chunking) - Error

Outstanding Points:
- I have tested the blosc compressors, but I haven't played with the level or shuffle yet
- I think the shuffle mode enums on HDF5 and NDPluginCodec are not in the same order, so that might need changing to allow validation of the NDArrays.
- I have not done any performance testing as I don't currently have an appropriate test server with a GPFS mount.

GDYendell commented:
Fixes #281 #349

GDYendell commented:
This doesn't build because ADSupport uses hdf5 1.10.1. In this version the method to use is `H5DOwrite_chunk`. In 1.10.3+, this is deprecated (and the build uses `H5_NO_DEPRECATED_SYMBOLS`) and `H5Dwrite_chunk` was added. We will need to require <1.10.3, require >=1.10.3, or put in some version checks.

ulrikpedersen commented:
Thanks @GDYendell.

@brunoseivam @MarkRivers @ajgdls: obviously still a few points left todo before this is complete (as Gary mentions above) so please don't merge just yet, but feel free to test and review the solution.

> In this version the method to use is H5DOwrite_chunk. In 1.10.3+, this is deprecated (and the build uses H5_NO_DEPRECATED_SYMBOLS) and H5Dwrite_chunk was added. We will need to require <1.10.3, require >=1.10.3, or put in some version checks.

We discussed this today and this should be solved quite easily with a macro...

ulrikpedersen commented:
> It looks like the code is requiring the user to configure the HDF5 plugin codec settings to match those of the incoming NDArrays?
> The HDF5 plugin gets the correct settings for things like data type, dimensions, etc. from the last previous array (LazyOpen=false) or from the first array to be written (LazyOpen=true). Why can't it just get the codec settings the same way?

Yes, that's right. This is to support the LazyOpen=false scenario. We didn't think it would be safe to depend on the last previous array settings. Although if we do depend on the last previous array, then it would arguably make the interface simpler and we could just ignore the chunk and compression parameters on the HDF5 plugin...?

> Why does it need to know the Blosc shuffle type, for example, when using DirectChunkWrite? That information is not needed to decode the chunk, because that information is in the Blosc header.

We need to set the parameters used to compress the NDArray in the HDF5 filter params (i.e. the `cd_values[]` array. This includes shuffle, compressor algorithm and level. The HDF5 library doesn't know about the blosc header (or any other compression library header) so must be told explicitly what parameters were set.

Also, in a more abstract view, I think the NDArray **should** contain all the information about how it was encoded in a generic fashion. The proposed struct covers all cases of compression that we're currently discussing.

ulrikpedersen commented:
The HDF5 library is told about the compression parameters so that it can mark the dataset as compressed and store some meta data about how it was compressed. In the case of Blosc, you're correct @MarkRivers  that it is not strictly required to decompress the data - but it is used to make the HDF5 file accurately describe the dataset. See for example the output of h5dump:

```
[up45@pc0009 testdata]$ h5dump -pH excalibur_26291_000001.h5 
HDF5 "excalibur_26291_000001.h5" {
GROUP "/" {
   DATASET "data" {
      DATATYPE  H5T_STD_U16LE
      DATASPACE  SIMPLE { ( 3025, 1536, 2048 ) / ( H5S_UNLIMITED, 1536, 2048 ) }
      STORAGE_LAYOUT {
         CHUNKED ( 1, 1536, 2048 )
         SIZE 251936364 (75.542:1 COMPRESSION)  <--  Nice level of compression achieved here :-)
      }
      FILTERS {
         USER_DEFINED_FILTER {
            FILTER_ID 32001                   <--  Blosc filter ID
            PARAMS { 2 2 2 6291456 4 1 1 }    <--- Compression parameters (cd_values[7])
         }
      }
      FILLVALUE {
         FILL_TIME H5D_FILL_TIME_IFSET
         VALUE  0
      }
      ALLOCATION_TIME {
         H5D_ALLOC_TIME_INCR
      }
   }
}
}
```

GDYendell commented:
As Ulrik said, I believe the compression parameters must be passed to set_filter so that reader applications can seamlessly open the datasets without the user even realising that it is compressed, as long as the have the filter library on the path.

I do like the idea of making it support LazyOpen, though. In the same way the chunk parameters are set automatically from the frame dimensions, it could also just copy the codec from the NDArray and use that to create the dataset and call H5Pset_filter.

ulrikpedersen commented:
> I do like the idea of making it support LazyOpen, though. In the same way the chunk parameters are set automatically from the frame dimensions, it could also just copy the codec from the NDArray and use that to create the dataset and call H5Pset_filter.

@GDYendell: I thought it did! Let's discuss - I'm coming over :smiley:

ulrikpedersen commented:
We had another look at this and @GDYendell is looking into using the previous array pArray[0] to check for compression settings on the incoming arrays and if the incoming NDArrays are pre-compressed, use those settings to override the HDF5 plugin compression parameters (to give feedback on _RBV so users can see what compression is actually used). Then lazyOpen should automatically do the right thing both when set to true and false as lazyOpen is fully implemented in the parent class of the HDF5 plugin...

MarkRivers commented:
> In the case of Blosc, you're correct @MarkRivers that it is not strictly required to decompress the data -

That is actually true of all of the existing supported NDArray codecs (Blosc, JPEG, lz4, bitshuffle/lz4).  ImageJ can decompress all of them from NTNDArrays.  The only information needed is the name of the codec, and the data type of the uncompressed data, not any of the compression parameters.

I agree that it is nice for the NDArray to have all of the compression information.  It appears that all of the compression parameters in HDF5 are required to be of type "int", correct?  

How should we handle compression parameters in NTNDArrays?  Currently codec.parameters contains only a single integer, which is the data type of the uncompressed data.  Should we also store the compression parameters there?

ulrikpedersen commented:
> That is actually true of all of the existing supported NDArray codecs (Blosc, JPEG, lz4, bitshuffle/lz4). ImageJ can decompress all of them from NTNDArrays. The only information needed is the name of the codec, and the data type of the uncompressed data, not any of the compression parameters.

Yes, that's probably correct. I'm concerned about the HDF5 Dynamically Loadable Filters - it is down to each of these filters to decide how to decode the compressed data from a HDF5 dataset and I wouldn't want to make the assumption that none of these hdf5 filters use the information stored in the HDF5 filter parameters.

> I agree that it is nice for the NDArray to have all of the compression information. It appears that all of the compression parameters in HDF5 are required to be of type "int", correct?

Yes, the [H5Pset_filter](https://support.hdfgroup.org/HDF5/doc/RM/RM_H5P.html#Property-SetFilter) function takes an unsigned int array and number of parameters as input.

> How should we handle compression parameters in NTNDArrays? Currently codec.parameters contains only a single integer, which is the data type of the uncompressed data. Should we also store the compression parameters there?

I haven't looked into NTNDArrays much, but yes, I would suggest we should code the same information in there - just don't know enough about it to say how. We should have enough information in the NTNDArray to completely regenerate an exact copy of the NDArray at the other end - and be able to store that in a HDF5 file too.

:christmas_tree: :santa: By the way, we are about to break for the holidays now. I think @GDYendell has already left and I'm off in about half an hour. We must continue with this in the new year  :fireworks: Merry Christmas! :christmas_tree:

GDYendell commented:
I have added macros to protect against using write chunk in <1.8.9 and to call the correct write_chunk method. I have also added hdf5_hl to the Makefile to use H5DOwrite_chunk for <1.10.3.

MarkRivers commented:
Are you ready for me to merge this?

ulrikpedersen commented:
> Are you ready for me to merge this?

@GDYendell ?

ajgdls commented on ADApp/pluginSrc/NDFileHDF5Dataset.cpp:
```diff
@@ -188,8 +254,43 @@ asynStatus NDFileHDF5Dataset::writeFile(NDArray *pArray, hid_t datatype, hid_t d
+    asynPrint(this->pAsynUser_, ASYN_TRACE_FLOW,
+              "%s::%s NDArray correctly chunked. Using direct chunk write\n",
+              fileName, functionName);
+    #if H5_VERSION_GE(1, 8, 11)
```
You test for H5 version greater than 1.8.11 here and then inside the if statement you test again with a #define.  That code can never be executed.

GDYendell commented:
@MarkRivers I think this is ready now.

MarkRivers commented:
I found 2 problems with this after I merged:

- NDFileHDF5.cpp I needed to add {} to avoid a compiler error
- commonDriverMakefile needs hdf5_hl library for direct chunk write (on some HDF5 versions).  It had been added to commonLibraryMakefile but not to commonDriverMakefile.  pluginTests failed to build.

I have made these changes and pushed.
